src/Compilers/CSharp/Portable/Operations/CSharpOperationNodes.cs

```diff
@@ -0,0 +1,1850 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.Operations
+{
+    internal sealed class CSharpLazyAddressOfOperation : LazyAddressOfOperation
+    {
+        private readonly CSharpOperationFactory _operationFactory;
```

Comment:
>private readonly CSharpOperationFactory _operationFactory [](start = 8, length = 57)

It feels like for lazy nodes we can also save on SemanticModel field, semantic model is available from the factory. For example, instead of having ```private readonly SemanticModel _owningSemanticModelOpt;``` in the base class, we could have ```protected TSemanticModelSource _modelSource``` where ```TSemanticModelSource : class, ISemanticModel```. Both CSharpOperationFactory and Semantic model would implement the interface, SemanticModel would simply return itself. Then, lazy nodes would use CSharpOperationFactory for TSemanticModelSource and early nodes would use SemanticModel for it.

Reply:
Meant to use ISemanticModel**Source** name rather than ISemanticModel for the interface name

---
In reply to: [232080038](https://github.com/dotnet/roslyn/pull/31010#discussion_r232080038) [](ancestors = 232080038)

Reply:
I took a look at what the implementation for this might look like. It would add a lot of generics in here, for the sake of saving a single pointer field on each type. It's an interesting idea, but I think it would add more complexity to the implementation than it would be worth.

---
In reply to: [232081732](https://github.com/dotnet/roslyn/pull/31010#discussion_r232081732) [](ancestors = 232081732,232080038)

Reply:
Filed [https://github.com/dotnet/roslyn/issues/31380](https://github.com/dotnet/roslyn/issues/31380) to track possibly doing this in the future.

---
In reply to: [232088619](https://github.com/dotnet/roslyn/pull/31010#discussion_r232088619) [](ancestors = 232088619,232081732,232080038)